pkg/api/api.go

```diff
@@ -778,183 +728,98 @@ func equalASCIIFold(s, t string) bool {
 	return s == t
 }
 
-// newStamperPutter returns either a storingStamperPutter or a pushStamperPutter
-// according to whether the upload is a deferred upload or not. in the case of
-// direct push to the network (default) a pushStamperPutter is returned.
-// returns a function to wait on the errorgroup in case of a pushing stamper putter.
-func (s *Service) newStamperPutter(r *http.Request) (storage.Storer, func() error, error) {
-	batch, err := requestPostageBatchId(r) // TODO: extrapolate the headers parsing to the handler level!
-	if err != nil {
-		return nil, noopWaitFn, fmt.Errorf("postage batch id: %w", err)
-	}
+type putterOptions struct {
+	BatchID  []byte
+	TagID    uint64
+	Deferred bool
+	Pin      bool
+}
+
+type putterSessionWrapper struct {
+	storer.PutterSession
+	stamper postage.Stamper
+	save    func() error
+}
 
-	deferred, err := requestDeferred(r) // TODO: extrapolate the headers parsing to the handler level!
+func (p *putterSessionWrapper) Put(ctx context.Context, chunk swarm.Chunk) error {
+	stamp, err := p.stamper.Stamp(chunk.Address())
 	if err != nil {
-		return nil, noopWaitFn, fmt.Errorf("request deferred: %w", err)
+		return err
 	}
+	return p.PutterSession.Put(ctx, chunk.WithStamp(stamp))
+}
 
-	if !deferred && s.beeMode == DevMode {
-		return nil, noopWaitFn, errUnsupportedDevNodeOperation
+func (p *putterSessionWrapper) Done(ref swarm.Address) error {
+	return multierror.Append(p.PutterSession.Done(ref), p.save()).ErrorOrNil()
```

Comment:
Since we are using Golang 1.20, I recommend using the `errors.Join` instead of the `multierror` package.
The same applies to the other places where the `multierror` is used.

Reply:
So for the validation package there is some formatting used. I dont want to add more changes here, so maybe we can do this in a separate change.